### PR TITLE
feat: (ui) Contract settings with Identity Naming

### PR DESCRIPTION
**Pull request by lordmallam**
NOTE: This PR is dependent on [AET-484](https://jira.ehealthafrica.org/browse/AET-484) being merged for less "files changed" 😄 
Implements the new contract settings mockup [here](https://jira.ehealthafrica.org/browse/AET-483)
Allowing custom naming for Identity Entity Type.
Stops default contract creation on new pipelines
Allows Identity mappings to be generated on non-identity contracts multiple times

@sabeen please help with styling the identity generation segment. 😃 

Related JIRA ticket: [AET-485](https://jira.ehealthafrica.org/browse/AET-485)

Unit Tests: WIP

### CODE REVIEW

File: aether-ui/aether/ui/assets/apps/pipeline/Pipeline.jsx [hunk 1]
```diff
@@ -31,7 +31,15 @@ import Mapping from './sections/Mapping'
...
-import { clearSelection, getPipelineById, selectContract, selectSection } from './redux'
+import {
+  clearSelection,
+  getPipelineById,
+  selectContract,
+  selectSection,
+  setExitWarning,
+  setEditing,
+  checkUnsavedContract
+} from './redux'
```
Comment (admbtlr) [inline aether-ui/aether/ui/assets/apps/pipeline/Pipeline.jsx:39]: Please order these lists alphabetically - it makes it easier to scan through them

File: aether-ui/aether/ui/assets/apps/pipeline/Pipeline.jsx [hunk 2]
```diff
@@ -51,28 +59,24 @@ class Pipeline extends Component {
...
   }
 
   componentDidUpdate (prevProps) {
     if (
-      (prevProps.section !== this.props.section ||
-      (prevProps.contract && this.props.contract && prevProps.contract.id !== this.props.contract.id)) &&
-      !this.state.isNew
+      (this.props.contract && (prevProps.section !== this.props.section ||
+      (prevProps.contract && prevProps.contract.id !== this.props.contract.id))) &&
+      !this.props.isNew
```
Comment (admbtlr) [inline aether-ui/aether/ui/assets/apps/pipeline/Pipeline.jsx]: I looked at the rest of the file to try and understand what's happening here, and I think you might have a bug on line 86: you reference `this.state.isNew`, but I don't think `isNew` ever gets put into state. Should that line reference `this.props.isNew` instead?

Comment (admbtlr) [reply aether-ui/aether/ui/assets/apps/pipeline/Pipeline.jsx]: And now I realise that there's no test coverage on this code... can we get some tests please?

File: aether-ui/aether/ui/assets/apps/pipeline/redux.jsx [hunk 3]
```diff
@@ -50,7 +50,11 @@ export const types = {
 
   CONTRACT_PUBLISH_PREFLIGHT: 'contract.publish.preflight',
   CONTRACT_PUBLISH_SUCCESS: 'contract.publish.success',
-  CONTRACT_PUBLISH_ERROR: 'contract.publish.error'
+  CONTRACT_PUBLISH_ERROR: 'contract.publish.error',
+  CONTRACT_SET_EDITING: 'contract_set_editing',
+  CONTRACT_EXIT_WARNING: 'contract_exit_warning',
+
+  SET_CALL_BACK: 'set_call_back'
```
Comment (admbtlr) [inline aether-ui/aether/ui/assets/apps/pipeline/redux.jsx]: I don't think it's a good idea to store a callback function in Redux - let's talk through some other options

File: aether-ui/aether/ui/assets/apps/pipeline/sections/Settings.jsx [hunk 4]
```diff
@@ -48,55 +50,114 @@ export class IdentityMapping extends Component {
...
+                  id='settings.identity.help-1'
+                  defaultMessage='Create an identity contract'
+                />
+              </label>
+            </div>
+            <p>
+              <FormattedMessage
+                id='settings.identity.help-2'
+                defaultMessage='An identity contract will produce entities
+                that are identical with the input. If you choose this setting.
```
Comment (admbtlr) [inline aether-ui/aether/ui/assets/apps/pipeline/sections/Settings.jsx]:
```suggestion
                that are identical with the input. If you choose this setting,
```

File: aether-ui/aether/ui/assets/apps/pipeline/sections/Settings.jsx [hunk 5]
```diff
@@ -48,55 +50,114 @@ export class IdentityMapping extends Component {
...
+                Aether will generate an Entity Type and Mapping for you.'
+              />
+            </p>
+            <p>
+              <FormattedMessage
+                id='settings.identity.help-3'
+                defaultMessage="This can be useful in situations where you
+                want to make use of Aether's functionality without transforming
+                  the data Alternatively, you can use the generate Entity Type
```
Comment (admbtlr) [inline aether-ui/aether/ui/assets/apps/pipeline/sections/Settings.jsx]:
```suggestion
                  the data. Alternatively, you can use the generate Entity Type
```

File: aether-ui/aether/ui/assets/apps/pipeline/sections/Settings.jsx [hunk 6]
```diff
@@ -48,55 +50,114 @@ export class IdentityMapping extends Component {
...
+            <h5>
+              <FormattedMessage
+                id='settings.identity.checked.help-1'
+                defaultMessage='This is an identity contract'
+              />
+            </h5>
+            <p>
+              <FormattedMessage
+                id='settings.identity.checked.help-2'
+                defaultMessage='All Entity Types and Mappings were generated automatically from the input'
```
Comment (admbtlr) [inline aether-ui/aether/ui/assets/apps/pipeline/sections/Settings.jsx]:
```suggestion
                defaultMessage='All Entity Types and Mappings were automatically generated from the input data'
```

**Review by admbtlr — CHANGES_REQUESTED** [review]

**Comment by sabeen** [comment]
hi @lordmallam! I tested again, and found that the "add contract" button from the pipelines overview doesn't open the "new contract" section but the input section of the pipeline.

**Comment by lordmallam** [comment]
> hi @lordmallam! I tested again, and found that the "add contract" button from the pipelines overview doesn't open the "new contract" section but the input section of the pipeline.

File: aether-ui/aether/ui/assets/apps/pipeline/sections/Settings.jsx [hunk 7]
```diff
@@ -48,55 +50,114 @@ export class IdentityMapping extends Component {
...
+                that are identical with the input. If you choose this setting,
+                Aether will generate an Entity Type and Mapping for you.'
+              />
+            </p>
+            <p>
+              <FormattedMessage
+                id='settings.identity.help-3'
+                defaultMessage="This can be useful in situations where you
+                want to make use of Aether's functionality without transforming
+                  the data. Alternatively, you can use the generate Entity Type
```
Comment (admbtlr) [inline aether-ui/aether/ui/assets/apps/pipeline/sections/Settings.jsx:94]: Indent

File: aether-ui/aether/ui/assets/apps/pipeline/Pipeline.jsx [hunk 8]
```diff
@@ -111,18 +115,52 @@ class Pipeline extends Component {
...
+          : <FormattedMessage id='pipeline.navbar.fullscreen' defaultMessage='fullscreen on' />
+        }
+      </div>
+    )
+
     return (
       <div className={`pipelines-container show-pipeline pipeline--${this.state.view}`}>
         { this.props.error && <ModalError error={this.props.error} /> }
-        <NavBar showBreadcrumb>
+        <NavBar showBreadcrumb cb={this.linksCallBack.bind(this)}>
```
Comment (admbtlr) [inline aether-ui/aether/ui/assets/apps/pipeline/Pipeline.jsx]: Please put all the bind calls into the constructor, e.g.
```
this.linksCallBack = this.linksCallBack.bind(this);
```

It makes the JSX more readable and removes duplication.

File: aether-ui/aether/ui/assets/apps/pipeline/Pipeline.jsx [hunk 9]
```diff
@@ -365,12 +416,12 @@ class Pipeline extends Component {
...
               defaultMessage='Mapping'
             />
-            { fullscreenDiv }
+
           </div>
 
           <div
             className='pipeline-nav-item__contracts'
-            onClick={() => { this.props.selectSection(CONTRACT_SECTION_ENTITY_TYPES) }}>
+            onClick={this.onContracts.bind(this)}>
```
Comment (admbtlr) [inline aether-ui/aether/ui/assets/apps/pipeline/Pipeline.jsx]: Here too

File: aether-ui/aether/ui/assets/apps/pipeline/sections/Settings.jsx [hunk 10]
```diff
@@ -222,31 +293,18 @@ class Settings extends Component {
...
           </div>
 
-          { showIdentityOption && <IdentityMapping {...this.props} /> }
+          { showIdentityOption && <IdentityMapping {...this.props}
+            onChange={(e) => this.setState({ isIdentity: e.target.checked })}
+            isIdentity={this.state.isIdentity}
+            showModal={this.state.showIdentityModal}
+            onModalToggle={(e) => this.setState({ showIdentityModal: e })}
+            contractName={this.state.contractName}
+            onSave={this.performSave.bind(this)} /> }
```
Comment (admbtlr) [inline aether-ui/aether/ui/assets/apps/pipeline/sections/Settings.jsx]: Here too

File: aether-ui/aether/ui/assets/apps/pipeline/sections/Settings.jsx [hunk 11]
```diff
@@ -266,14 +324,27 @@ class Settings extends Component {
...
-                defaultMessage='Close'
+                id='settings.button.cancel'
+                defaultMessage='Cancel'
               />
             </span>
           </button>
+          { !contract.is_read_only &&
+            <button
+              className='btn btn-d btn-primary btn-big ml-4'
+              onClick={this.onSave.bind(this, contract)}>
```
Comment (admbtlr) [inline aether-ui/aether/ui/assets/apps/pipeline/sections/Settings.jsx]: And here...

**Review by admbtlr — CHANGES_REQUESTED** [review]
Much better!

**Review by admbtlr — APPROVED** [review]

**Pull request merged**